docs/design/coreclr/jit/DeabstractionAndConditionalEscapeAnalysis.md

```diff
@@ -0,0 +1,800 @@
+
+(14) Interestingly there are [vestiges of an empty-list static enumerator](https://github.com/dotnet/runtime/blame/5ae9467d6ab7b31bf1cda35ca15ca5a2d21d3046/src/libraries/System.Private.CoreLib/src/System/Collections/Generic/List.cs#L1187), so presumably at one point there was a static enumerator instance like we hae for the empty array case, but we realized that empty list enumeration could share the empty array instance. See [dotnet/runtime#82499)](https://github.com/dotnet/runtime/pull/82499).
+
+(15) Aside from explicit looping, the JIT is able to transform recursive tail calls into loops, though this can only be done when the live part of the method state is the arguments to the call. Note thanks to inlining and GDV this can kick in in cases where the method is part of a recursive cycle of calls.
+
+(16) We could (and likely should)  try to handle both, but we need to be thoughtful about how much cloning this entails. A naive version would end up creating 4 copies of the inner foreach, likely we don't want to have the version where we have the fast outer path and slow inner path, unless saving a heap allocation ends up being super-valuable.
+
+(17) Inlining methods with EH is on the .NET 10 roadmap, and perhaps this is one case to use as motivation.
+
+More broadly, we should consider having the inliner aggressively inline if a call site can be passed a locally allocated object. Here it is tricky because the connection between object allocation and call is indirect, and we don't do data flow during inlining. Now, if we did escape analysis before inlining, we could perhaps target the methods that cause the allocations to escape, but we'd then need to update / redo the analysis for each inline so we can handle the transitive cases... (this is where some kind of up-front hinting would be very useful, so the JIT would know which inlines could end up preventing escape).
+
+(18) Overall this seems like something we could improve on. The `MoveNext` call site must dominate the `get_Current` site, so surely the class profile for the one likely can be applied to the other. But we have limited abilities to make such inferences today, as all this is done very early in the jit. And (as here) if `get_Current` does not postdominate `MoveNext` we can't be sure that the conditional logic in between is doing some computation that might alter the profile. But when the dominating profile is monomorphic then projecting it to the dominated site is probably a safe bet.
+
+(19) Note this is an instance where inlining a "cold" method is beneficial, as it touches the same state as a hot path. We have seen these before in other contexts.
```

Comment:
Seems like some footnotes are not referenced e.g. (19). I think Markdown has some built-in ways for footnotes? (I am fine with it as is, nice doc!)

Reply:
Hmm, I renumbered at one point some maybe something got mixed up.

Reply:
Found it.. there were two refs to 18.